### DISCUSSION TITLE
Variables

### DISCUSSION THREAD

**Discussion by RickCogley** [retypeapp/retype#213] [Q&A] [answered]
Hi, are there any variables besides "year" available? Searched but didn't find, so I appreciate any pointer.

**Comment by geoffreymcgill** [comment, marked as answer]
No other system variables are available at this time, but this is functionality we plan to expand heavily in the future.

We have content `partials` under development. Once that is in place, adding system variables and even custom user defined variables will be possible.

**Reply by RickCogley** [reply]
Thanks, that'll be cool

**Comment by geoffreymcgill** [comment]
The new partials and variables support is going to be amazing. A full content templating language is being integrated. Functions for arrays including looping, string manipulation, math, regex, date, etc.

You will be able to define global custom variables in your `retype.yml` or at a page level in the page metadata. Arrays of custom object variables can be defined too. Variables can be passed into partials or other components. Partials can be nested. Global variables can be used to set configs in the page metadata, including the `key` names.

---

The following demonstrates configuring a simple partial with an `alert` component whose variant is set in a custom global variable and the text message is passed into the partial from a custom variable set in the page metadata:

#### retype.yml
```
data:
  customVariant: success
```

#### _includes/custom-alert.md

```md
!!!{{ customVariant }} {{ $.title | string.capitalize }}
{{ $.text }}
!!!
```

#### sample.md

```
---
data:
  customText: This is custom alert message defined in a variable.
  customTitle: Holy cow!
---
# Sample

{{ include "custom-alert" text:customText title:customTitle }}
```

The sample above would render as:

<img width="796" alt="Screenshot 2023-06-14 at 7 21 17 AM" src="https://github.com/retypeapp/retype/assets/62210/cac96db1-07e7-4383-ab30-ca86bc867c07">

---

This new functionality is going to unlock a whole new level of content authoring flexibility. Essentially programmable content.

Will be interesting to see how authors will leverage this new functionality.

The plan is to include in Retype `v1.12.0`.

**Reply by DannyBen** [reply]
This looks great, and coming pretty soon I see, which is even nicer.

Have you decided where the partials live? A dedicated directory (e.g. `components`)? Allowing subdirectories (e.g. `components/alerts/danger.md`)?

**Reply by geoffreymcgill** [reply]
> Have you decided where the partials live?

Not yet, although it will be very flexible.

**Reply by RickCogley** [reply]
That's cool. One thing that I have seen is personalization of document sets. For instance, you have a service that gives you an url like `myco.theirservice.io` and say their API docs reference that exact thing, instead of a generic one. For example: 

Use the `https://myco.theirservice.io/api/v2/flangeblurp` endpoint to enumerate your flanges that contain blurps. 

I assume this is usually done server side, but it's still intriguing.

**Reply by geoffreymcgill** [reply]
> One thing that I have seen is personalization of document sets

@RickCogley err, you kinda lost me there. Can you rephrase that feature?

**Reply by RickCogley** [reply]
My musing out loud is confusing; sorry about that. 

Not sure if it's a feature that makes sense in a static site, now that I think of it, but what I'm remembering is how Fortrabbit's docs change when you're logged in. The highlighted bit in the screenshot is specific to our account. It's pretty slick that it updates like that, but I assume it's server side and being handled by pulling in variables from our account after we log in. 

<img width="1144" alt="image" src="https://user-images.githubusercontent.com/512328/144530210-191d32d2-eec3-4e45-8755-3995e274226d.png">

As for a static site, using variables might let us easily personalize the content of the site's pages, to include information for a specific audience at build time. It might also be cool to be able to pass those variables on the command line, so you could also do the personalization at build time, in say a CI scenario.

**Reply by geoffreymcgill** [reply]
@RickCogley Thanks for the clarification. I've been thinking of similar personalization features. Eventually, I think we can make this happen, although it will take a while until we get to that point.

> It might also be cool to be able to pass those variables on the command line, so you could also do the personalization at build time, in say a CI scenario.

Excellent idea. We hadn't thought of this. We can definitely build this.

**Reply by geoffreymcgill** [reply]
Passing custom variables in the command line might already be possible using [`--override`](https://retype.com/guides/cli/#usage-of---override), although I haven't tested yet.

We could also add a specific `--data` option. We will keep this in mind as we build out the feature.

Setting var values using Environment variables is also being investigated.

**Reply by DannyBen** [reply]
> Setting var values using Environment variables is also being investigated.

I was about to suggest it. This will be great.

**Reply by geoffreymcgill** [reply]
The new templating (includes, variables, functions, conditionals, looping, etc) functionality is finally complete, and it's awesome.

This was a massive piece of work that touched every aspect of the Retype engine and took waa-a-a-a-aay longer than we had hoped, but it's now basically ready to go. We still want to do another in depth round of testing before releasing a preview.

As always, performance was the priority. We were able to make performance improvements throughout Retype in addition to layering on the new templating engine, and Retype is still screaming fast.

The underlying templating engine is [Scriban](https://github.com/scriban/scriban/tree/master/doc) which supports [Liquid](https://shopify.github.io/liquid/) style syntax. Everything (plus more) that is supported by Scriban and Liquid is now supported within Retype. It's an extremely powerful and flexible templating system.

We still have the task of documenting the new templating functionality, although just because of the sheer enormity of what is now possible, that documentation is going to have to roll out over some time. We'll get the basics documented and some advanced samples, then link off to the [Scriban](https://github.com/scriban/scriban/tree/master/doc) docs for the full API.

Please keep in mind that is for content templating only, not the website _wrapper_ template. We're hoping to support complete website templating in a future release.

**Reply by geoffreymcgill** [reply]
Templating is now available in the Retype [`v2.0.0-preview`](https://github.com/retypeapp/retype/discussions/287#discussioncomment-2321442) release. If you have some time, we would certainly appreciate your feedback on the preview release before we go live with `v2.0.0` (likely next week).

**Reply by amarcolini** [reply]
Is there now any documentation for this feature? It looks really cool and I want to give it a try, but I'm not sure how to get started.

**Comment by dipbhi** [comment]
@geoffreymcgill,
Just tried using it exactly as your sample, but for me it's printing `{{ custom-alert text="{{ custom-text }}" title="{{ custom-title }}" }}` on the page instead of showing the alert defined in the `custom-alert` partial.

Using retype version 2.4.0.

Are the partials supposed to be in a specific directory? 

Noticed these two warnings in the console as well:

```
WARNING: [xyz.md:8,17] Error while parsing assign expression: Expression is only allowed for a top level assignment in: <target_expression> = <value_expression>
WARNING: [xyz.md:8,42] Error while parsing assign expression: Expression is only allowed for a top level assignment in: <target_expression> = <value_expression>
````

**Reply by geoffreymcgill** [reply]
I just ran across this old templating sample and corrected to use the final release syntax, see

https://github.com/retypeapp/retype/discussions/213#discussioncomment-1737438
